### PR TITLE
Feature/custom panel add viz

### PR DESCRIPTION
**Pull request by ps48**
### Description
Added features to add new visualization, save visualization states, save visualization data query and added time range support

### Issues Resolved
Save Visualization States - #92 
Save Visuailzation Data Query - #93 
Add Multiple Visualizations - #76 
Add Time Range Support - #89 
Add new Visualization Basic - #73 

~~NOTE: After merging this branch: we need to change OpenSearch-Dashbaord dependencies to the following version in package.json: 
"react-grid-layout": "^1.2.5",
"@types/react-grid-layout": "^1.1.2",~~

~~Post changing don't forget to rebuild with  `yarn osd bootstrap`~~

EDIT: The library was reverted back to the original version as used in OpenSearch-Dashbaord. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,17 +12,21 @@
...
   "devDependencies": {
+    "@types/react-grid-layout": "^1.1.2",
     "@types/enzyme-adapter-react-16": "^1.0.6",
     "@types/react-plotly.js": "^2.2.4",
     "@types/react-test-renderer": "^16.9.1",
     "cypress": "^5.0.0",
     "enzyme-adapter-react-16": "^1.15.2",
     "eslint": "^6.8.0",
     "jest-dom": "^4.0.0"
+    "react-graph-vis": "^1.0.5",
```
Comment (joshuali925) [inline package.json:29]: why is this moved to `devDependencies`?

File: package.json [hunk 2]
```diff
@@ -12,17 +12,21 @@
...
+    "react-plotly.js": "^2.5.1"
```
Comment (joshuali925) [inline package.json:30]: this is already added in `dependencies`?

File: common/constants/custom_panels.ts [hunk 3]
```diff
@@ -9,7 +9,21 @@
  * GitHub history for details.
  */
 
-export const CUSTOM_PANELS_API_PREFIX = '/api/observability/custom_panels';
+export const CUSTOM_PANELS_API_PREFIX = '/api/observability/operational_panels';
 export const CUSTOM_PANELS_DOCUMENTATION_URL = 'https://www.opensearch.org';
 export const CREATE_PANEL_MESSAGE = 'Enter a name to describe the purpose of this custom panel.';
-export const DATE_FORMAT = 'MM/DD/YYYY hh:mm A';
+export const DATE_FORMAT = 'YYYY-MM-DD HH:mm:ss';
+export const INDEX_REGEX = /(search source|source|index)\s*=\s*([^|\s]+)/i;
```
Comment (joshuali925) [inline common/constants/custom_panels.ts:16]: this is also used by Eric, maybe move this to shared and change his function to use it as well?

File: package-lock.json [hunk 4]
```diff
@@ -0,0 +1,1984 @@
+{
```
Comment (joshuali925) [inline package-lock.json:1]: i don't think we need package-lock.json, could you remove this file?

File: public/components/custom_panels/custom_panel_view.tsx [hunk 5]
```diff
@@ -9,37 +9,342 @@
...
+    setEnd(end);
+    setRecentlyUsedRanges(
+      recentlyUsedRange.length > 10 ? recentlyUsedRange.slice(0, 9) : recentlyUsedRange
+    );
+    setIsLoading(true);
+    startLoading();
+  };
+
+  const startLoading = () => {
+    setTimeout(stopLoading, 1000);
```
Comment (joshuali925) [inline public/components/custom_panels/custom_panel_view.tsx:111]: why is this hard coded?

File: public/components/custom_panels/forms/add_new_visualizations.tsx [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+ * setPanelVisualizations: Setter for panelVisualizations object
+ * setToast: Create Toast function 
+ */
+
+type Props = {
+  closeVizWindow: () => void;
+  pplService: PPLService;
+  panelVisualizations: VisualizationType[];
+  setPanelVisualizations: React.Dispatch<React.SetStateAction<VisualizationType[]>>;
+  setToast: (title: string, color?: string, text?: string) => void;
```
Comment (joshuali925) [inline public/components/custom_panels/forms/add_new_visualizations.tsx:44]: FYI might be better to use `notifications` in `CoreStart` if possible, so that core dashboards handles toasts

File: public/components/custom_panels/forms/add_new_visualizations.tsx [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+            <EuiLink href="https://opensearch.org/docs/search-plugins/ppl/index/" target="_blank">
+              Learn More
+            </EuiLink>{' '}
+          </Fragment>
+        }
+        fullWidth={true}
+      >
+        <EuiTextArea
+          placeholder="Placeholder text"
+          aria-label="Use aria labels when no actual label is in use"
```
Comment (joshuali925) [inline public/components/custom_panels/forms/add_new_visualizations.tsx:218]: should these be something like "PPL query"?

File: common/constants/custom_panels.ts [hunk 8]
```diff
@@ -9,7 +9,21 @@
  * GitHub history for details.
  */
 
-export const CUSTOM_PANELS_API_PREFIX = '/api/observability/custom_panels';
+export const CUSTOM_PANELS_API_PREFIX = '/api/observability/operational_panels';
 export const CUSTOM_PANELS_DOCUMENTATION_URL = 'https://www.opensearch.org';
 export const CREATE_PANEL_MESSAGE = 'Enter a name to describe the purpose of this custom panel.';
-export const DATE_FORMAT = 'MM/DD/YYYY hh:mm A';
+export const DATE_FORMAT = 'YYYY-MM-DD HH:mm:ss';
```
Comment (joshuali925) [inline common/constants/custom_panels.ts:15]: i don't think UI displays date with year month date, is this used by PPL? if so could you add a comment or add "PPL" in the variable name?

File: public/components/custom_panels/custom_panel_view.tsx [hunk 9]
```diff
@@ -9,37 +9,342 @@
...
+                  value={filterBarValue}
+                  fullWidth={true}
+                  onChange={(e) => onChange(e)}
+                  disabled={inputDisabled}
+                />
+              </EuiFlexItem>
+              <EuiFlexItem grow={false}>
+                <EuiSuperDatePicker
+                  isLoading={isLoading}
+                  dateFormat="MM/DD/YYYY hh:mm:ss A"
```
Comment (joshuali925) [inline public/components/custom_panels/custom_panel_view.tsx:298]: could this be a constant?

File: public/components/custom_panels/forms/add_saved_visualizations.tsx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    });
+    recentlyUsedRange.unshift({ start, end });
+    setStart(start);
+    setEnd(end);
+    setRecentlyUsedRanges(
+      recentlyUsedRange.length > 10 ? recentlyUsedRange.slice(0, 9) : recentlyUsedRange
+    );
+    setIsLoading(true);
+    startLoading();
+    console.log('time change', start, end);
```
Comment (joshuali925) [inline public/components/custom_panels/forms/add_saved_visualizations.tsx:62]: could you remove the console.log? also this `onTimeChange` looks identical in other places, is it possible to move it to utils.tsx to reduce duplicates and for easier unit tests?

File: public/components/custom_panels/helpers/utils.tsx [hunk 11]
```diff
@@ -9,6 +9,120 @@
...
+  endTime: string,
+  panelFilterQuery: string
+) => {
+  const indexMatchArray = originalQuery.match(INDEX_REGEX);
+  if (indexMatchArray == null) {
+    throw Error('index not found in Query');
+  }
+  const indexPartOfQuery = indexMatchArray[0];
+  const filterPartOfQuery = originalQuery.replace(INDEX_REGEX, '');
+  const timeQueryFilter = ` | where ${timestampField} > timestamp('${convertDateTime(
```
Comment (joshuali925) [inline public/components/custom_panels/helpers/utils.tsx:56]: probably won't make a difference in most cases, but should we make the time range inclusive (e.g. `>=` and `<=`)?

File: public/components/custom_panels/home.tsx [hunk 12]
```diff
@@ -153,59 +163,85 @@ export const Home = ({ http, chrome, parentBreadcrumb }: Props) => {
...
+              customPanels={customPanelData}
+              createCustomPanel={createCustomPanel}
+              setBreadcrumbs={chrome.setBreadcrumbs}
+              parentBreadcrumb={parentBreadcrumb}
+              renameCustomPanel={renameCustomPanel}
+              cloneCustomPanel={cloneCustomPanel}
+              deleteCustomPanel={deleteCustomPanel}
+              setToast={setToast}
+            />,
+            4
```
Comment (joshuali925) [inline public/components/custom_panels/home.tsx:224]: i removed the requirement to put index, could you pull latest commits and try removing the second argument

File: package-lock.json (hunk 4 above)
Comment (ps48) [reply package-lock.json:1]: Will remove this 👍🏽

File: package.json [hunk 13]
```diff
@@ -12,17 +12,21 @@
   },
   "dependencies": {
     "@reduxjs/toolkit": "^1.6.1",
+    "@types/react-plotly.js": "^2.5.0",
     "plotly.js-dist": "^2.2.0",
-    "react-graph-vis": "^1.0.5",
+    "react-grid-layout": "^1.2.5",
```
[the remark this answers is not in the document]
Comment (ps48) [reply package.json:17]: The version in dashboards was giving me issue while using latest documented react components. I tried using two different versions on dashboard and plugin, it surely gives build error due to version conflict. I'll retry this once again with older version mentioned in dashboards, if not then I'll create an issue with bump up.

File: package.json (hunk 1 above)
Comment (ps48) [reply package.json:29]: This got created while resolving merge conflicts. Will move it back.

File: package.json (hunk 2 above)
Comment (ps48) [reply package.json:30]: same as above! 👍🏽

File: common/constants/custom_panels.ts (hunk 3 above)
Comment (ps48) [reply common/constants/custom_panels.ts:16]: Will move it to this file https://github.com/opensearch-project/trace-analytics/blob/feature/observability/common/constants/shared.ts later.

File: common/constants/custom_panels.ts (hunk 8 above)
Comment (ps48) [reply common/constants/custom_panels.ts:15]: Yes this is the PPL format will edit the name and move it to shared constants.

File: public/components/custom_panels/custom_panel_view.tsx (hunk 9 above)
Comment (ps48) [reply public/components/custom_panels/custom_panel_view.tsx:298]: Will add it shared it to shared constants, so other modules can use it.

File: public/components/custom_panels/forms/add_saved_visualizations.tsx (hunk 10 above)
Comment (ps48) [reply public/components/custom_panels/forms/add_saved_visualizations.tsx:62]: Indeed this is duplicating code. Will move this to utils.

File: public/components/custom_panels/custom_panel_view.tsx (hunk 5 above)
Comment (ps48) [reply public/components/custom_panels/custom_panel_view.tsx:111]: This was supposed to be a placeholder. But not needed now I guess. Will remove it.

File: public/components/custom_panels/forms/add_new_visualizations.tsx (hunk 6 above)
Comment (ps48) [reply public/components/custom_panels/forms/add_new_visualizations.tsx:44]: The toasts still works without `notifications` being added. Could you please explain a bit more on how adding it would help? If you could point me to a doc that would be great.

File: public/components/custom_panels/forms/add_new_visualizations.tsx (hunk 7 above)
Comment (ps48) [reply public/components/custom_panels/forms/add_new_visualizations.tsx:218]: I missed this. Thanks

File: public/components/custom_panels/helpers/utils.tsx (hunk 11 above)
Comment (ps48) [reply public/components/custom_panels/helpers/utils.tsx:56]: I can make this inclusive on both ends `>= and <=`. Let's make this consistent across modules in observability. Looping in Eric. @mengweieric what's your opinion on this?

File: public/components/custom_panels/home.tsx (hunk 12 above)
Comment (ps48) [reply public/components/custom_panels/home.tsx:224]: Will rebase and then make this change.

File: public/components/custom_panels/sub_components/visualization_panel.tsx [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  toTime?: string;
+};
+
+export const VisualizationPanel = ({
+  visualizationId,
+  visualizationTitle,
+  fromTime,
+  toTime,
+}: Props) => {
+  const [isPopoverOpen1, setIsPopoverOpen1] = useState(false);
```
Comment (mengweieric) [inline public/components/custom_panels/sub_components/visualization_panel.tsx]: Just curious for why these variables were named with an 1 in the end?

File: public/components/custom_panels/forms/add_new_visualizations.tsx (hunk 6 above)
Comment (joshuali925) [reply public/components/custom_panels/forms/add_new_visualizations.tsx:44]: Not sure if there are docs, I'm referring to [these functions](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/notifications/toasts/toasts_api.tsx#L172-L262) from `coreStart.notifications.toasts`.

This makes toasts less flexible since we are not defining the `EuiToast` component, but I can see some benefits:
1. we don't have to maintain toast states and the separate toast component here:
https://github.com/opensearch-project/trace-analytics/blob/d6b3bc98ab333f5e5346ca366d9fb5e4f360c2c1/public/components/custom_panels/home.tsx#L200-L206
2. using `addDanger`, `addSuccess` and other functions takes care of toast color and icon, making it more uniform with Dashboards core UI
3. if there is a toast sent by core or other plugins, toasts sent by `custom_panels` will not be placed on top and cover it up
4. toasts will not disappear if user navigates away from `custom_panels` to other pages like event_explorer or trace_analytics, this can be good or bad depending on the use case

The current implementation also works, so this is just a FYI note

Comment (ps48) [reply public/components/custom_panels/forms/add_new_visualizations.tsx:44]: I was thinking of a common toast function implementation for all observability modules for long term. The functions form `coreStart.notifications.toasts` look great  and they solve common toast issue. In pt.3 by cover it up, if you mean overlap of toasts then we should implement it using  `notifications`.  But, pt.4 is a bit tricky and may be UX team can help us more on it. For now let's keep it as is. 

If you are starting to implement toasts for trace analytics, we can probably have a second opinion from the UX team, that would save you some double effort.

Comment (joshuali925) [reply public/components/custom_panels/forms/add_new_visualizations.tsx:44]: Trace analytics is read only, so I don't think toasts were needed. For point 3 yes, I have not tested it myself but I saw this behavior in reporting (reporting doesn't have access to `coreStart`): if there's an existing toast from core and reporting sends a toast, user cannot see the toast from core unless they close the reporting toast. we can check with UX team on whether we should persist toasts across pages

File: public/components/custom_panels/sub_components/plot_sample.tsx [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (ps48) [reply public/components/custom_panels/sub_components/plot_sample.tsx]: I did use this from your template for event explorer visualizations. I'll just reuse it and pass layout changes as parameters.

File: public/components/custom_panels/sub_components/add_visualization.tsx [hunk 16]
```diff
@@ -117,19 +169,113 @@ export const AddVizView = ({ closeVizWindow }: Props) => {
...
-        <EuiRadioGroup
-          options={radios}
-          idSelected={radioIdSelected}
-          onChange={(id) => onChangeRadio(id)}
-          name="radio group"
-          aria-aria-rowcount={1}
-        />
-      </EuiFormRow>
-      {radioIdSelected == 'radio1' ? (
+  // const formBody = (
```
[the remark this answers is not in the document]
Comment (ps48) [reply public/components/custom_panels/sub_components/add_visualization.tsx]: That's from an older commit I guess.

File: public/components/custom_panels/helpers/utils.tsx (hunk 11 above)
Comment (mengweieric) [reply public/components/custom_panels/helpers/utils.tsx]: Yes, inclusive makes sense here.

File: public/components/custom_panels/sub_components/visualization_panel.tsx (hunk 14 above)
Comment (ps48) [reply public/components/custom_panels/sub_components/visualization_panel.tsx]: I used this from EUI examples. Probably stuck with the same variable name, will change this.

File: package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (ps48) [reply package.json]: Tried again, reverted back the older version. It seems to work now. My older implementation had some issues. Hadn't checked again with this new implementation using Responsive Grid Layout Component. So, no need to bump version I guess.

File: package.json [hunk 17]
```diff
@@ -13,12 +13,12 @@
   "dependencies": {
     "@reduxjs/toolkit": "^1.6.1",
     "plotly.js-dist": "^2.2.0",
-    "react-grid-layout": "^1.2.5",
+    "react-grid-layout": "^0.16.2",
```
Comment (joshuali925) [inline package.json:16]: you can remove this and the `@types/react-grid-layout` if it's the same version defined in Dashboards' package.json

**Review by joshuali925 — APPROVED** [review]

File: package.json (hunk 17 above)
Comment (ps48) [reply package.json:16]: Removed 👍🏽

**Review by mengweieric — APPROVED** [review]
